### PR TITLE
Immutably and implicitly borrow all pattern ids for their guards (NLL only)

### PR DESCRIPTION
**Pull request by pnkfelix**
This is an important piece of rust-lang/rust#27282.

It applies only to NLL mode. It is a change to MIR codegen that is currently toggled on only when NLL is turned on. It thus affect MIR-borrowck but not the earlier static analyses (such as the type checker).

This change makes it so that any pattern bindings of type T for a match arm will map to a `&T` within the context of the guard expression for that arm, but will continue to map to a `T` in the context of the arm body.

To avoid surfacing this type distinction in the user source code (which would be a severe change to the language and would also require far more revision to the compiler internals), any occurrence of such an identifier in the guard expression will automatically get a deref op applied to it.

So an input like:
```rust
let place = (1, Foo::new());
match place {
  (1, foo) if inspect(foo) => feed(foo), 
  ...  
}
```
will be treated as if it were really something like:
 ```rust
let place = (1, Foo::new());
match place {
    (1, Foo { .. }) if { let tmp1 = &place.1; inspect(*tmp1) }
                    => { let tmp2 = place.1; feed(tmp2) },
    ...
}
```

And an input like:
```rust
let place = (2, Foo::new());
match place {
    (2, ref mut foo) if inspect(foo) => feed(foo),
    ... 
}
```
will be treated as if it were really something like:

```rust
let place = (2, Foo::new());
match place {
    (2, Foo { .. }) if { let tmp1 = & &mut place.1; inspect(*tmp1) }
                    => { let tmp2 = &mut place.1; feed(tmp2) },
    ... 
}
```

In short, any pattern binding will always look like *some* kind of `&T` within the guard at least in terms of how the MIR-borrowck views it, and this will ensure that guard expressions cannot mutate their the match inputs via such bindings. (It also ensures that guard expressions can at most *copy* values from such bindings; non-Copy things cannot be moved via these pattern bindings in guard expressions, since one cannot move out of a `&T`.)

### CODE REVIEW

**Comment by pnkfelix** [comment]
cc @arielb1  (normally I wouldn't ping someone who I think is quite busy with other parts of their life, but I've seen ariel chiming in on other PR's and I think he may be interested in doing so on this one.)

File: src/librustc_mir/build/block.rs [hunk 1]
```diff
@@ -113,7 +114,8 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                     // Declare the bindings, which may create a visibility scope.
                     let remainder_span = remainder_scope.span(this.hir.tcx(),
                                                               &this.hir.region_scope_tree);
-                    let scope = this.declare_bindings(None, remainder_span, lint_level, &pattern);
+                    let scope = this.declare_bindings(None, remainder_span, lint_level, &pattern,
+                                                      false);
```
Comment (nikomatsakis) [inline src/librustc_mir/build/block.rs]: Nit: a plain `false` like this is always a bit unclear to me as a reader. Maybe we could put a newtype here, like `declare_bindings(..., InGuard(false))` (and `struct InGuard(bool)`)?

File: src/librustc_mir/build/matches/mod.rs [hunk 2]
```diff
@@ -793,47 +924,130 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+                    // While we could have kept track of this with a
+                    // flag or collection of such bindings, the
+                    // treatment of all of these cases is uniform, so
+                    // we should be safe just avoiding the code
+                    // without maintaining such state.)
+                    let local_for_arm_body = self.storage_live_binding(
+                        block, binding.var_id, binding.span, OutsideGuard);
+                    self.schedule_drop_for_binding(binding.var_id, binding.span, OutsideGuard);
+                    let rvalue = Rvalue::Ref(region, borrow_kind, binding.source.clone());
+                    self.cfg.push_assign(block, source_info, &local_for_arm_body, rvalue);
```
Comment (nikomatsakis) [inline src/librustc_mir/build/matches/mod.rs]: So before I thought we were going to need some sort of cast here -- I was trying to remember why -- I believe it was looking forward to the point where we try to maintain a shared borrow the value being matched over the whole match, in which case this first `&mut` borrow would be illegal...

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Did a first pass -- looking good -- but we need some tests, I suppose? (Or maybe I just didn't get to that commit, actually)

File: src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// Here below is a case that needs to keep working: we only use the
+// binding via immutable-borrow in the guard, and we move into the arm
+// body.
+fn allow_move_into_arm_body() {
+    match Some(&4) {
+        None => {},
+        mut foo if foo.is_some() && false => { foo.take(); () }
+        Some(s) => std::process::exit(*s),
```
Comment (nikomatsakis) [inline src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs]: never mind I see test now :)

File: src/librustc_mir/build/matches/mod.rs (hunk 2 above)
Comment (pnkfelix) [reply src/librustc_mir/build/matches/mod.rs]: okay let me look into this, see if I can make a test case that exposes the scenario you describe.

File: src/librustc_mir/build/block.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc_mir/build/block.rs]: Or I should probably just reuse the `enum` I introduced somewhere else in this PR...

Comment (pnkfelix) [reply src/librustc_mir/build/block.rs]: (well, technically this boolean flag has a different meaning currently than the `WithinGuard`/`OutsideGuard` enum... but I'll figure something out. I too dislike plain booleans.)

**Comment by the8472** [comment]
Does this solve https://github.com/rust-lang/rfcs/issues/2036?

**Comment by pietroalbini** [comment]
Ping from triage! Can @nikomatsakis (or someone else from @rust-lang/compiler) review this?

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me once nits are addressed

File: src/librustc_mir/build/matches/mod.rs [hunk 4]
```diff
@@ -770,21 +783,139 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+        // currently encoded in a pretty awkward fashion; namely, by
+        // passing a boolean to bind_matched_candidate_for_arm_body
+        // indicating whether all of the by-ref bindings were already
+        // initialized.
+        //
+        // * Also: pnkfelix thinks "laziness" is natural; but since
+        //   MIR-borrowck did not complain with earlier (universally
+        //   eager) MIR codegen, laziness might not be *necessary*.
+
+        let autoref = self.hir.tcx().nll();
```
Comment (arielb1) [inline src/librustc_mir/build/matches/mod.rs]: Why do we generate different MIR in NLL and non-NLL modes?

File: src/librustc_mir/build/matches/mod.rs [hunk 5]
```diff
@@ -770,21 +783,139 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
         if let Some(guard) = candidate.guard {
```
Comment (arielb1) [inline src/librustc_mir/build/matches/mod.rs]: Sorry for being late with this, but I feel that the "guard bit" mapping - figuring out whether a local should access the "arm" version or the dereference of the "guard" version - belongs more in HAIR than in MIR.

File: src/librustc_mir/build/matches/mod.rs (hunk 2 above)
Comment (arielb1) [reply src/librustc_mir/build/matches/mod.rs]: > I believe it was looking forward to the point where we try to maintain a shared borrow the value being matched over the whole match, in which case this first `&mut` borrow would be illegal...

Isn't the answer here is "this is why we have 2-phase borrows"? There's this cute example, which compiles today, is not a soundness issue, and works in the 2phi-borrow interpretation, but does not work if 2phi is not used correctly.

```Rust
use std::cell::Cell;

fn main() {
    let mut b = 4;
    let a : Cell<Option<&u32>> = Cell::new(None);
    match b {
        ref mut btag if { a.set(Some(&*btag)); true } => {
            // `btag` is immutably borrowed here.
            println!("a={:?} b={}", a, btag);
        }
        _ => {}
    }
}
```

File: src/librustc_mir/build/matches/mod.rs (hunk 4 above)
Comment (nikomatsakis) [reply src/librustc_mir/build/matches/mod.rs]: That was my assumption -- that said, thinking more about this, if we are going to try and merge NLL down into a warning period, we are going to need the MIR generation to be the same both with and without NLL enabled.

File: src/librustc_mir/build/matches/mod.rs (hunk 5 above)
Comment (nikomatsakis) [reply src/librustc_mir/build/matches/mod.rs]: Hmm. I don't have a strong opinion, but that seems reasonable. How do you envision it working? I could imagine having e.g. distinct HIR nodes for "reference variable from guard" or something...?

File: src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+// except according to those terms.
+
+//revisions: ast mir
+//[mir] compile-flags: -Z borrowck=mir -Z nll
+
+#![feature(rustc_attrs)]
+
+// Here is arielb1's basic example from rust-lang/rust#27282
+// that AST borrowck is flummoxed by:
+
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs]: Yes, this was not intended to fix *everything*.

File: src/librustc_mir/build/matches/mod.rs (hunk 5 above)
Comment (arielb1) [reply src/librustc_mir/build/matches/mod.rs]: What I was thinking of was having variables in the HAIR refer to an enum
```Rust
enum HairVar {
    Binding(ast::NodeId),
    GuardRef(ast::NodeId)
}
```

MIR construction does not quite care about the exact `NodeId` of variables, only the scope of their definitions etc., so this should work out.

Comment (arielb1) [reply src/librustc_mir/build/matches/mod.rs]: > distinct HIR nodes for "reference variable from guard"

The above option would mean that we put it right in HAIR construction. Of course, we could also put it in resolution (using 2 separate `NodeId`s), but I understand the implications of that less.

Comment (nikomatsakis) [reply src/librustc_mir/build/matches/mod.rs]: (In case it wasn't clear, I meant "distinct HAIR nodes", I edited my comment -- but it sounds like we were thinking of similar things)

Comment (pnkfelix) [reply src/librustc_mir/build/matches/mod.rs]: (at this point I would prefer to land this PR in the short term and come back later to the question of whether the guard bit mapping should part of the HAIR rather than the MIR... I will open an issue for that task once this has landed though...)

**Comment by pnkfelix** [comment]
okay I've addressed the crucial feedback from arielb1, and I *think* I've done so in a way that does not perturb the PR too significantly... so I think I'll re-add niko's r+ once this passes travis... (famous last words...)

File: src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (pnkfelix) [reply src/test/compile-fail/nll/issue-27282-guards-always-borrow.rs]: Just an FYI: I agree that the this did not put in the full fix to the core problem. However, the example outlined by @arielb1 is *not* accepted by the compiler (under NLL) once this PR is in place. In particular, you get these errors:

```
error[E0382]: use of moved value: `*v`
  --> ../../ariel-example-2-27282.rs:11:9
   |
8  |             (|| { let bar = v; bar.take() })();
   |             -------------------------------- value moved here
...
11 |         Some(s) => std::process::exit(*s),
   |         ^^^^^^^ value used here after move

error[E0382]: use of moved value: `v.0`
  --> ../../ariel-example-2-27282.rs:11:14
   |
8  |             (|| { let bar = v; bar.take() })();
   |             -------------------------------- value moved here
...
11 |         Some(s) => std::process::exit(*s),
   |              ^ value used here after move
   |
   = note: move occurs because `v.0` has type `&i32`, which does not implement the `Copy` trait
```

**Comment by pnkfelix** [comment]
I am now realizing that this change is probably *very* relevant to [RFC 107](https://github.com/rust-lang/rfcs/blob/master/text/0107-pattern-guards-with-bind-by-move.md) i.e. rust-lang/rfcs#107.
